### DISCUSSION TITLE
Authelia with nginx login loops TOO_MANY_REDIRECTS

### DISCUSSION THREAD

**Discussion by crebollobr** [authelia/authelia#6753] [Q&A] [answered]
I installed authelia and nginx in bare metall server
I cofigure then with documentantion on site https://www.authelia.com/integration/proxies/nginx/

authelia log:

DEBU[1547] No matching rule for subject username= groups= ip=172.27.48.1 and url https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/favicon.ico (method GET) applying default policy

nginx log:

2024/02/28 13:38:04 [error] 1462#1462: *611 auth request unexpected status: 431 while sending to client, client: 172.27.48.1, server: , request: "GET /?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/?rd=https://teste.com.br/ HTTP/1.1", host: "teste.com.br"
[config.txt](https://github.com/authelia/authelia/files/14437531/config.txt)
[default.txt](https://github.com/authelia/authelia/files/14437532/default.txt)

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the logs you've provided, it appears that you're encountering an issue with a redirection loop (`?rd=` parameter being appended multiple times) and an HTTP 431 error from NGINX, which indicates that the server is unwilling to process the request because its header fields are too large.

The redirection loop might be caused by misconfiguration in how Authelia and NGINX handle the redirection and authentication requests. Specifically, the repeated appending of the `?rd=` parameter suggests that the request is being redirected back to Authelia multiple times without ever being authenticated successfully or reaching the intended destination.

The HTTP 431 error suggests that the accumulation of redirection parameters (`?rd=`) in the URL has caused the request headers to become too large for NGINX to process.

To address these issues, consider the following steps:

1. **Review Redirection Configuration**: Ensure that the redirection to the Authelia authentication portal is configured correctly in NGINX. This typically involves using the `error_page` directive to redirect unauthenticated requests to Authelia. Make sure that this redirection is not causing an infinite loop by inadvertently redirecting authenticated requests or by misinterpreting Authelia's responses.

2. **Check Authelia Configuration**: Verify that Authelia is correctly configured to handle authentication requests and that it's properly integrated with NGINX. This includes checking the `auth_request` directive in your NGINX configuration, which should point to an internal location that proxies to Authelia's `/api/verify` endpoint.

3. **Examine Request Headers**: The HTTP 431 error indicates that the request headers are too large. This could be due to the redirection loop appending too many parameters to the URL. Reducing the number of redirections or the size of the request headers may help. You might also want to increase the `large_client_header_buffers` setting in NGINX to allow for larger headers if necessary.

4. **Simplify the Test Case**: If possible, simplify your configuration to a minimal setup that reproduces the issue. This can help isolate the problem and make it easier to identify the misconfiguration.

5. **Consult Documentation and Community Resources**: Review the Authelia and NGINX documentation for guidance on integrating Authelia with NGINX. The Authelia community, including forums and discussion boards, can also be a valuable resource for troubleshooting specific configuration issues.

Without access to the specific details of your configuration files (`config.txt` and `default.txt` mentioned in your logs), these suggestions are somewhat general. If you need more detailed assistance, consider sharing relevant portions of your NGINX and Authelia configuration files (with sensitive information redacted).

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
I'd probably suggest including the nginx configs, seems like something is not configured per the guide.

**Comment by crebollobr** [comment]
I have configured with examples of documentation.
Degging, if i acess https://teste.com.br:9091/ i get a cookie
if i acess https://teste.com.br/ i have no cookie

**Comment by james-d-elliott** [comment, marked as answer]
Okay, seems like you didn't configure it like the guide as the guide doesn't configure the server block that reverse proxies to authelia with the auth_request directive. Fix that as per the portal example.

**Comment by crebollobr** [comment]
I copy and paste the configuration on portal.
/config/nginx/snippets/authelia-authrequest.conf
/config/nginx/snippets/authelia-location.conf
https://www.authelia.com/integration/proxies/nginx/


[config.txt](https://github.com/authelia/authelia/files/14447942/config.txt)
[default.txt](https://github.com/authelia/authelia/files/14447943/default.txt)

**Reply by james-d-elliott** [reply]
You have to do all of them, authelia portal, alll 3 snippets, and the protected app. The example is used in every change made to authelia for automated testing. It works.. You're giving me everything but what I asked for, I'm starting to lose my patience / interest.

I want to help but without the nginx configs (authelia portal and protected app) it's just not practical. There should be a file with nearly exactly this content serving your authelia portal if you followed the guide (and this is where you redirect users to):

```nginx
server {
    listen 80;
    server_name auth.*;

    return 301 https://$server_name$request_uri;
}

server {
    listen 443 ssl http2;
    server_name auth.*;

    include /config/nginx/snippets/ssl.conf;

    set $upstream http://authelia:9091;

    location / {
        include /config/nginx/snippets/proxy.conf;
        proxy_pass $upstream;
    }

    location /api/verify {
        proxy_pass $upstream;
    }
}
```

**Reply by crebollobr** [reply]
Its worked. Thanks

**Comment by crebollobr** [comment]
Log of authelia

TRAC[0296] Request hit                                   method=GET path=/api/verify remote_ip=172.27.48.1
TRAC[0296] Headers=GET /api/verify HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0
Host: teste.com.br:9091
X-Original-Url: https://teste.com.br/
X-Original-Method: GET
X-Forwarded-Method: GET
X-Forwarded-Proto: https
X-Forwarded-Host: teste.com.br
X-Forwarded-Uri: /
X-Forwarded-For: 172.27.48.1
Sec-Ch-Ua: "Chromium";v="122", "Not(A:Brand";v="24", "Microsoft Edge";v="122"
Sec-Ch-Ua-Mobile: ?0
Sec-Ch-Ua-Platform: "Windows"
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Fetch-Dest: document
Accept-Encoding: gzip, deflate, br
Accept-Language: pt-BR,pt;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6
  method=GET path=/api/verify remote_ip=172.27.48.1

**Comment by crebollobr** [comment]
Log of authelia

behind proxy 

root@note360:~# TRAC[1579] Request hit                                   method=GET path=/api/verify remote_ip=172.27.55.60
TRAC[1579] Headers=GET /api/verify HTTP/1.1
User-Agent: curl/7.81.0
Host: teste.com.br:9091
X-Original-Url: https://teste.com.br/
X-Original-Method: GET
X-Forwarded-Method: GET
X-Forwarded-Proto: https
X-Forwarded-Host: teste.com.br
X-Forwarded-Uri: /
X-Forwarded-For: 172.27.55.60
Accept: */*
  method=GET path=/api/verify remote_ip=172.27.55.60
TRAC[1579] Using X-Original-URL header content as targeted site URL  method=GET path=/api/verify remote_ip=172.27.55.60
TRAC[1579] Checking if we need check the authentication backend for an updated profile for .  method=GET path=/api/verify remote_ip=172.27.55.60
DEBU[1579] Check authorization of subject username= groups= ip=172.27.55.60 and object https://teste.com.br/ (method GET).
DEBU[1579] No matching rule for subject username= groups= ip=172.27.55.60 and url https://teste.com.br/ (method GET) applying default policy
INFO[1579] Access to https://teste.com.br/ (method GET) is not authorized to user <anonymous>, responding with status code 401  method=GET path=/api/verify remote_ip=172.27.55.60
TRAC[1579] Replied (status=401)                          method=GET path=/api/verify remote_ip=172.27.55.60

root@note360:~#
root@note360:~#
root@note360:~#
root@note360:~#
root@note360:~#

direct

root@note360:~# TRAC[1627] Request hit                                   method=GET path=/ remote_ip=172.27.55.60
TRAC[1627] Replied (status=200)                          method=GET path=/ remote_ip=172.27.55.60
